ambry-frontend/src/main/java/com/github/ambry/frontend/s3/S3ListHandler.java

```diff
@@ -119,4 +140,58 @@ private ReadableStreamChannel serializeAsXml(RestRequest restRequest, Page<Named
     xmlMapper.writeValue(outputStream, result);
     return new ByteBufferReadableStreamChannel(ByteBuffer.wrap(outputStream.toByteArray()));
   }
+
+  private class CallbackChain {
```

Comment:
I suspect that this isn't an isolated case, I wonder whether it makes sense to make this generic to handle all features we don't yet support and share among all handlers?

Reply:
I also have the same thought. 
But the problem is that for HTTP request, we don't know the request type until the time the code is passing the parameters.
In the handleGet, we have the code as below, besides the multipart code, we have two condition,
1. the path has blob name, it goes to the "else". 
2. the path doesn't have the blob name, it goes to the s3ListHandler.

If we want to create a standalone s3OtherHandlers, we need a way to know what are the "other handles". 
Let me know if you have any suggestions.

```
      } else if (isS3Request(restRequest)) {
        if (S3BaseHandler.isListBucketRequest(restRequest)) {
          s3ListHandler.handle(restRequest, restResponseChannel,
              (result, exception) -> submitResponse(restRequest, restResponseChannel, result, exception));
        } else if (S3BaseHandler.isMultipartListPartRequest(restRequest)) {
          s3MultipartUploadHandler.handle(restRequest, restResponseChannel,
              (result, exception) -> submitResponse(restRequest, restResponseChannel, result, exception));
        } else {
          getBlobHandler.handle(requestPath, restRequest, restResponseChannel,
              (result, exception) -> submitResponse(restRequest, restResponseChannel, result, exception));
        }
      } else if (requestPath.matchesOperation(Operations.NAMED_BLOB)
```

Reply:
I didn't mean to create an other S3 handler, but just the CallbackChain class as a utility that can be shared, so that we don't have to repeat the same CallbackChain class repeatedly.

Reply:
Just wrap the CallbackChain is a little bit weird to me.
I'm thinking creating one S3GetHandler and wrap all the s3 handlers together. Let me work on it first and discuss more after the change.

Reply:
Thanks!